### ISSUE TITLE
Possible use of uninitialized GPU memory

### ISSUE THREAD

**Issue by f0k** [Theano/Theano#4220]
A Lasagne user found a problem with nondeterministic results for a convolutional autoencoder (discussion at https://groups.google.com/d/msg/lasagne-users/bfPPuQBBzmM/TatY1IuZBAAJ).

He provided a demonstration script: https://gist.github.com/sveitser/73984476c8bf3794e3f3
It should be completely deterministic, but when run on GPU with cuDNN, it provides different results at each run, and it is influenced by allow_gc and lib.cnmem. My suspicion is that it reads uninitialized memory somewhere.

I'm reproducing a slightly simplified version of that script here -- unfortunately, it requires Lasagne (bleeding-edge, not 0.1):
```python
from lasagne import layers, objectives, updates
import numpy as np
import theano
import theano.tensor as T

np.random.seed(0)

input_var = T.tensor4('X')
target_var = T.tensor4('X2')

input = layers.InputLayer((None, 1, 5, 5), input_var)
conv = layers.Conv2DLayer(input, num_filters=64, filter_size=(3, 3), pad=1,
                          nonlinearity=None, b=None)
pool = layers.MaxPool2DLayer(conv, pool_size=(3, 3), stride=(2, 2))
inverse_pool = layers.InverseLayer(pool, pool)
network = layers.InverseLayer(inverse_pool, conv)

prediction = layers.get_output(network)
loss = objectives.squared_error(prediction, target_var).mean()

g = theano.grad(loss, input_var)
f = theano.function([input_var, target_var], {'out': prediction, 'grad': g})
X = np.random.randn(1, 1, 5, 5).astype(np.float32)

print(f(X, X))
print(f(X, X))
print(f(X, X))
```
An InverseLayer(A, B) takes its input from layer A and passes it through the gradient implementation of layer B (obtained via theano.grad).

Running this gives some very weird behaviour:
```
THEANO_FLAGS=device=gpu1,allow_gc=1,lib.cnmem=0 python grad2.py 
Using gpu device 1: Tesla K40c (CNMeM is disabled, CuDNN 4004)
{'grad': array([[[[-0.0044742 ,         nan,         nan,         nan,  0.00493809],
         [ 0.09747709,         nan,         nan,         nan,  0.02103304],
         [-0.00183207,         nan,         nan,         nan,  0.08208349],
         [ 0.02717943,         nan,         nan,         nan,  0.05635365],
         [ 0.00623315,         nan,         nan,         nan,  0.01645008]]]], dtype=float32), 'out': array([[[[ 0.15524647,  0.06967667, -0.08527909, -0.58669513, -0.10754481],
         [ 0.07445988, -1.15406203,  1.43113112,  0.52614611,  0.10764439],
         [ 0.2656078 ,  2.15410089, -0.62238789, -2.6701932 ,  0.25219753],
         [ 0.05570559, -0.86656028, -0.84023333, -0.56997377,  0.09399533],
         [-0.1621647 ,  1.02166164, -0.49901724, -0.42257324, -0.46175131]]]], dtype=float32)}
{'grad': array([[[[-0.0044742 ,  0.08431669, -0.13045976,  0.05704633,  0.00493809],
         [ 0.09747709, -0.14178798,  0.6584478 , -0.02132855,  0.02103304],
         [-0.00183207,  0.59277028, -0.21802083, -0.80688292,  0.08208349],
         [ 0.02717943, -0.60644859, -0.07882077, -0.25750688,  0.05635365],
         [ 0.00623315,  0.2853801 , -0.05407535,  0.05283685,  0.01645008]]]], dtype=float32), 'out': array([[[[ 0.15524647,  0.06967667, -0.08527909, -0.58669513, -0.10754481],
         [ 0.07445988, -1.15406203,  1.43113112,  0.52614611,  0.10764439],
         [ 0.2656078 ,  2.15410089, -0.62238789, -2.6701932 ,  0.25219753],
         [ 0.05570559, -0.86656028, -0.84023333, -0.56997377,  0.09399533],
         [-0.1621647 ,  1.02166164, -0.49901724, -0.42257324, -0.46175131]]]], dtype=float32)}
{'grad': array([[[[-0.0044742 ,  0.08431669, -0.13045976,  0.05704633,  0.00493809],
         [ 0.09747709, -0.14178798,  0.6584478 , -0.02132855,  0.02103304],
         [-0.00183207,  0.59277028, -0.21802083, -0.80688292,  0.08208349],
         [ 0.02717943, -0.60644859, -0.07882077, -0.25750688,  0.05635365],
         [ 0.00623315,  0.2853801 , -0.05407535,  0.05283685,  0.01645008]]]], dtype=float32), 'out': array([[[[ 0.15524647,  0.06967667, -0.08527909, -0.58669513, -0.10754481],
         [ 0.07445988, -1.15406203,  1.43113112,  0.52614611,  0.10764439],
         [ 0.2656078 ,  2.15410089, -0.62238795, -2.6701932 ,  0.25219753],
         [ 0.05570559, -0.86656028, -0.84023333, -0.56997377,  0.09399534],
         [-0.1621647 ,  1.02166164, -0.49901724, -0.42257324, -0.46175131]]]], dtype=float32)}
```
Notice how the network output is the same for the three identical function calls, but the first gradient is different from the other two. Running this script directly again gives the same results. On the other GPU, results differ, but are also consistent across runs:
```
$ THEANO_FLAGS=device=gpu0,allow_gc=1,lib.cnmem=0 python grad2.py 
Using gpu device 0: Tesla K40c (CNMeM is disabled, CuDNN 4004)
/raid/user/jan/install/Theano-git/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")
{'grad': array([[[[ -4.47420217e-03,              nan,              nan,
                       nan,   4.93808696e-03],
         [             nan,              nan,              nan,
                       nan,  -1.35928233e+35],
         [             nan,              nan,              nan,
                       nan,   5.14289453e+35],
         [             nan,              nan,              nan,
                       nan,   5.43827653e+35],
         [  6.23314548e-03,              nan,              nan,
                       nan,   1.64500773e-02]]]], dtype=float32), 'out': array([[[[ 0.15524647,  0.06967667, -0.08527909, -0.58669513, -0.10754481],
         [ 0.07445988, -1.15406203,  1.43113112,  0.52614611,  0.10764439],
         [ 0.2656078 ,  2.15410089, -0.62238789, -2.6701932 ,  0.25219753],
         [ 0.05570559, -0.86656028, -0.84023333, -0.56997377,  0.09399534],
         [-0.1621647 ,  1.02166164, -0.49901724, -0.42257324, -0.46175131]]]], dtype=float32)}
{'grad': array([[[[ -4.47420217e-03,   2.97598037e+34,  -6.99805152e+33,
           -2.42000392e+34,   4.93808696e-03],
         [             nan,              nan,   4.24068294e+34,
            5.77924325e+35,  -1.35928233e+35],
         [             nan,              nan,  -8.88560282e+34,
           -1.11101145e+36,   5.14289453e+35],
         [             nan,              nan,  -1.10108654e+36,
            1.13418846e+36,   5.43827653e+35],
         [  6.23314548e-03,   1.24295447e+36,  -1.48858833e+35,
            5.77501445e+35,   1.64500773e-02]]]], dtype=float32), 'out': array([[[[ 0.15524647,  0.06967667, -0.08527909, -0.58669513, -0.10754481],
         [ 0.07445988, -1.15406203,  1.43113112,  0.52614611,  0.10764439],
         [ 0.2656078 ,  2.15410089, -0.62238789, -2.6701932 ,  0.25219753],
         [ 0.05570559, -0.86656028, -0.84023333, -0.56997377,  0.09399534],
         [-0.1621647 ,  1.02166164, -0.49901724, -0.42257324, -0.46175131]]]], dtype=float32)}
...
```
When activating lib.cnmem or setting allow_gc=1, I sometimes get the (supposedly) correct result right from the start:
```
$ THEANO_FLAGS=device=gpu1,allow_gc=1,lib.cnmem=500 python grad2.py 
Using gpu device 1: Tesla K40c (CNMeM is enabled with initial size: 500 MB, CuDNN 4004)
{'grad': array([[[[-0.0044742 ,  0.0446834 , -0.04515738, -0.07043385,  0.00493809],
         [ 0.24721681, -0.65398145,  1.33470106,  0.37110272,  0.15801281],
         [-0.02761342,  1.21112263, -0.46286565, -1.5341841 ,  0.1575903 ],
         [ 0.02101405, -0.89249873, -0.44807422, -0.49744055,  0.13981481],
         [ 0.00623315,  0.34763888, -0.10060667,  0.05908062,  0.01645008]]]], dtype=float32), 'out': array([[[[ 0.15524647,  0.06967667, -0.08527909, -0.58669513, -0.10754481],
         [ 0.07445988, -1.15406203,  1.43113112,  0.52614611,  0.10764439],
         [ 0.2656078 ,  2.15410089, -0.62238795, -2.6701932 ,  0.25219753],
         [ 0.05570559, -0.86656028, -0.84023333, -0.56997377,  0.09399533],
         [-0.1621647 ,  1.02166164, -0.49901724, -0.42257324, -0.46175131]]]], dtype=float32)}
{'grad': array([[[[-0.0044742 ,  0.0446834 , -0.04515738, -0.07043385,  0.00493809],
...
```
Or some random other result, but that one also consistently:
```
$ THEANO_FLAGS=device=gpu1,allow_gc=1,lib.cnmem=50 python grad2.py 
Using gpu device 1: Tesla K40c (CNMeM is enabled with initial size: 50 MB, CuDNN 4004)
/raid/user/jan/install/Theano-git/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")
{'grad': array([[[[ -4.47420217e-03,   2.05365270e-01,  -9.28192735e-02,
           -1.19270615e-01,   4.93808696e-03],
         [  1.67613000e-01,  -6.87860250e-01,   1.09213769e+00,
            1.28836423e-01,   7.60879293e-02],
         [ -4.24346216e-02,   1.35460758e+00,  -1.45263314e-01,
           -1.58063817e+00,   2.71790534e-01],
         [  8.58022198e-02,  -9.31332588e-01,  -3.52701306e-01,
           -5.55166423e-01,   1.94965497e-01],
         [  6.23314548e-03,   4.33806717e-01,  -1.58974260e-01,
            1.19394495e-03,   1.64500773e-02]]]], dtype=float32), 'out': array([[[[ 0.15524647,  0.06967667, -0.08527909, -0.58669513, -0.10754481],
         [ 0.07445988, -1.15406203,  1.43113112,  0.52614611,  0.10764439],
         [ 0.2656078 ,  2.15410089, -0.62238795, -2.6701932 ,  0.25219753],
         [ 0.05570559, -0.86656028, -0.84023333, -0.56997377,  0.09399533],
         [-0.1621647 ,  1.02166164, -0.49901724, -0.42257324, -0.46175131]]]], dtype=float32)}
...
```

Running with optimizer_excluding=cudnn gets rid of all these surprises. So does removing the pooling layer and its inverse.

So it seems there's either something wrong in cuDNN, or in the way Theano's cuDNN-max-pooling gradient and gradient-gradient Ops are implemented. Maybe something about the pooling switches (IIRC, cuDNN remembers them for the backward pass), but that's a wild guess. Any idea?

**Comment by lamblin** [comment]
The first thing I would try in order to get rid of some non-deterministic behaviour would be to set the config flags `dnn.conv.algo_bwd_filter` and `dnn.conv.algo_bwd_data` to `'deterministic'`, since the inverse conv layer is not deterministic by default in CuDNN (it uses `AtomicAdd`).

When running with an instance of `NanGuardMode` as the compilation mode, are you able to pinpoint where the nan appears first?

Does it work with older versions of CuDNN? With the released version of CuDNN v4 (4007)? Maybe older versions of Theano?

**Comment by nouiz** [comment]
I'm able to reproduce it.

Using the flag optimizer_including=alloc_empty_to_zeros fix the problem
(after a few crash fix inside Theano, I'll make a PR later about that)

If I use mode=DebugMode and work around the false positive due to
PyCObject, I get this optimization that raise the error:
local_gpu_downsample_factor_max_grad_grad

I continue to check a little, but maybe I won't find other thing tonigh.

**Comment by nouiz** [comment]
I have a fix:

https://github.com/Theano/Theano/pull/4222

the 2nd derivative max pool on the GPU don't support strides and we where
still moving it to the GPU.

**Comment by f0k** [comment]
Nice catch! Thanks for describing some of the debugging steps as well, they're good to know.
